### PR TITLE
test: make E2E tests faster using HAR files

### LINKED ISSUE #435: Tests failing after 00:00
Some of the tests are consistently failing after 00:00, impacting the reliability of our testing suite

This is most likely due to the API not having information about the next day buses yet, I suggest modifying the tests to check for that case

### LINKED ISSUE #454: make e2e tests faster using har files
first problem: e2e tests are slow because they send real queries to the backend
second problem: #435 

solution: use HAR files to capture backend responses

example: https://github.com/hasadna/open-bus-map-search/blob/5cad578b5f8a4cb9c697cd45f21ecea3af7a3b8e/tests/dashboard.spec.ts#L6

### PR DESCRIPTION
**Pull request by R894**
# Description
### Summary of changes
1. Add HAR files for every test using ```advancedRouteFromHAR```
2. Add ```setBrowserTime``` function to ```utils.ts```
3. Add ```setFakeTime``` function to ```BasePage``` (relying on ```setBrowserTime``` utility)
4. Add ```getPastDate``` function to ```utils.ts``` which returns a date from the past
5. Move custom matcher logic seperate function for better code readability

### Concerns
HAR files take a lot of space (around 13mb for each one on average), there could be a better solution than the one proposed here, I would love to hear your thoughts

closes #435  
closes #454

### CODE REVIEW

File: tests/utils.ts [hunk 1]
```diff
@@ -37,10 +38,44 @@ export const test = baseTest.extend({
...
-  return formattedDate
+export function getPastDate(): Date {
+  return moment('2024-02-12 15:00:00').toDate()
+}
+
+export async function setBrowserTime(date: Date, page: Page | BrowserContext) {
+  const fakeNow = date.valueOf()
+
+  // Update the Date accordingly
+  await page.addInitScript(`{
```
Comment (NoamGaash) [inline tests/utils.ts:49]: suggestion - you can pass a function instead of a string to enjoy linting and intellisense

File: tests/utils.ts [hunk 2]
```diff
@@ -37,10 +38,44 @@ export const test = baseTest.extend({
...
+      const fieldsToRemove = ['t', 'date_from', 'date_to']
+      ;[a, b] = [a, b].map((url) => {
+        const urlObj = new URL(url)
+        fieldsToRemove.forEach((field) => urlObj.searchParams.delete(field))
+        return urlObj.toString()
+      })
+      return a === b
+    },
+  })
 }
```
Comment (NoamGaash) [inline tests/utils.ts:79]: just a suggestion - you can create a single instance of the `urlMatcher`

```suggestion
export const urlMatcher: Matcher = customMatcher({
    urlComparator(a, b) {
      const fieldsToRemove = ['t', 'date_from', 'date_to']
      ;[a, b] = [a, b].map((url) => {
        const urlObj = new URL(url)
        fieldsToRemove.forEach((field) => urlObj.searchParams.delete(field))
        return urlObj.toString()
      })
      return a === b
    },
  })
```

And on usage:
```
advancedRouteFromHAR('tests/HAR/clearbutton.har', {
      updateContent: 'embed',
      update: false,
      notFound: 'abort',
      url: /stride-api/,
      matcher: urlMatcher, // NOTE - no `()` are needed
})
```

**Review by NoamGaash — APPROVED** [review]
It's perfect!
I made some suggestions with my thoughts, but feel free to merge it as-is if you prefer (those changes are not required, your code seems great just as it is)

**Comment by NoamGaash** [comment]
regarding your concern - I think you're right, the HAR file has too many redundant information. I made this issue, I hope I'll fix it someday:
https://github.com/NoamGaash/playwright-advanced-har/issues/13

File: tests/utils.ts (hunk 2 above)
Comment (R894) [reply tests/utils.ts]: updated, thanks!

File: tests/utils.ts (hunk 1 above)
Comment (R894) [reply tests/utils.ts]: updated!

**Pull request merged**